Serialization fixes for x.39

jeff303 opened the pull request:
This will be a combined PR for a sequence of fixes for serialization issues to be included in 0.39.

# Multi commit review (experimental)

GitHub only allows a PR per branch.  This means that even splitting logical work across different commits is not enough.  Those commits need to be on different branches, and the base branch of successive PRs needs to be set to that of the parent commit.  This becomes very cumbersome, especially as changes are made and multiple branches have to be rebased.

This is an attempt to use the existing GitHub PR flow to review individual commits, but still on a single branch, to reduce that burden.

## Review Process
The full details of each distinct change (a commit) will be kept in the respective commit messages.  Review each commit separately, using the GitHub task list functionality:

Once an individual commit is completely reviewed, check its box.  If changes are needed, leave comments on the respective lines as normal, then request changes.  Once all boxes are checked, mark the overall PR as approved.

- [ ] Commit 1: support loading linked cards in different collections (fixes: #15812)
- [ ] Commit 2: support dashboard card click actions (fixes: #14566)
- [ ] Commit 3: stop nesting cards and check for unresolved query dependencies (fixes: #14914)
- [ ] Commit 4: logging improvements
- [ ] Commit 5: handle collection namespaces and retry pulses (fixes: #14910)
- [ ] Commit 6: handle dimension entry within template-tags
- [ ] Commit 7: handle textbox cards
- [ ] Commit 8: handle dimension parameter mappings (fixes: #15971)
- [ ] Commit 9: handle field literals correctly
- [ ] Commit 10: accept only-db-ids option for dump 
- [ ] Commit 11: don't include personal collections of other users (fixes: #15120)
- [ ] Commit 12: handle native query snippets (fixes: #14909)
- [ ] Commit 13: handle source-table for joins (fixes #15283)
- [ ] Commit 14: cleanup and refactoring
- [ ] Commit 15: preserve temporal-unit in source-field clause
- [ ] Commit 16: misc cleanup and fixes
- [ ] Commit 17: handle joining to cards (fixes #15283 for real, hopefully this time)

jeff303 committed:
Serialization fixes: support dashboard card click actions

Architectural changes:
Add visualization settings shared model namespace with utility functions

Adding tests for viz settings

Changes to serialization:
Update visualization_settings in dump to convert column_settings keys from field ID to fully qualified name, and update the click action target with the fully qualified name
Also converting targetId in top level click action to fully qualified name

Doing the reverse in load, using the same utility code

Changes to serialization tests:
Updating the existing dashcard in the "world" roundtrip test to include the click action

Adding assertion for the click action in the assert-loaded-entity implementation for Dashboard

jeff303 committed:
Serialization fixes: logging improvements

Include causes when logging errors to aid with diagnosis

Add BEGIN and END messages for both dump and load commands

Add dynamic var to suppress printing of name lookup error on first pass load

jeff303 commented:
Commit 1 should fix #14914

Commit 2 should fix #14566

jeff303 committed:
Serialization fixes: stop nesting cards and check for unresolved query dependencies

In names.clj, stop qualifying a card's name by its dataset query's source card ID (this is what caused the "nested" directory structure)

In load.clj, start checking that either the source query [:dataset_query :query] is resolved properly as a card ID or table ID, and if neither, then propagate an ::unresolved-names entry up into the card (causing it to be reloaded on second pass)

Fix logging of unresolved name info

Test changes:
Add two new cards to the "world", one that lives in the root collection but depends on a card within a collection for its query, and one opposite (i.e. lives in a collection but depends on a root card for its query)

Adding new check that collection names match after loading

jeff303 committed:
Serialization fixes: handle collection namespaces and retry pulses

Architectural changes:
Considering a collection namespace to be part of its fully qualified name (putting in a subdirectory called :namespace when writing to disk)
Accounting for that change when dumping and loading, and upsert (i.e. namespace is part of unique columns)
Bumping serialization format version to 2 because of this

Changes to load:
Add load-pulses fn that works similarly to others, to facilitate reloading
Add similar logic as other entities to reload a pulse if the :card_id is not found

Model changes:
Add hack fn to compare namespace values for the "namespace does not change" assertion with appropriate comment

Test changes:
Added test for pulse with reload

jeff303 committed:
Serialization fixes: support loading linked cards in different collections

Architectural changes:
Lay foundation for more generic retry mechanism
Establish a keyword to represent a failed name->id lookup that can happen during load, and map that to information about which part of the card resolution failed (for logging purposes)
Similar to dashboards, add a separate fn for performing a load of cards
Invoking the load-cards fn from the load multimethod implementation when some cards failed name resolution

Test enhancements:
Add new multimethod to perform more specific assertions on the loaded entities
For cards - check that the native SQL of the loaded card matches that of the original (pre-dump) instance
For dashboards - check that the linked card for each dashboard card series matches

Bug fixes:
Apply the mbql-id->fully-qualified-name fn on serialize to filter, breakout, and aggregation clauses
Updating "My Test" card to include filter clause so that fix is exercised
Instead of upserting users with a "changeme" password, simply do not attempt to change the password
Adding :engine to the identity condition of Database (so unrelated test-data instances don't get trashed on load)

Cleanup:
Remove unused specs namespace

jeff303 committed:
Serialization fixes: handle dimension entry within template-tags

Add :field to the recognized clauses under mbql-entity-reference? (since that is how field refs are now represented in MBQL)

Add temp-field helper fn in test_util.clj to reduce some of the "world" verbosity

Adding new test card to world that uses template-tags with a field reference under dimension

jeff303 committed:
Serialization fixes: handle textbox cards

Add support for passing textbox (virtual) cards through visualization_settings.cljc

Updating load test to add a textbox type card to the dashboard and ensuring it is persisted

jeff303 committed:
Serialization fixes: handle dimension parameter mappings

In visualization_settings.cljc:
Adding handling for parameterMapping
Adding new roundtrip test for parameter mappings
Some refactoring to split up larger fns

In serialize.clj, convert all parts of the map that contain the dimension key to use fully qualified names for the respective fields

In load.clj, do the opposite (convert back to new IDs)

Updating test to add parameter mapping to an existing click action, and adding new assertions for that to ensure it is preserved

Fixing serialize test now that the "ids" can themselves be maps (i.e within dimension vector)

In visualization_settings_test.cljc
Adding new roundtrip test for problem dashboard

jeff303 committed:
Serialization fixes: handle field literals correctly

On load, only attempt to translate field name to ID if it's actually a fully qualified field name

Adding new fn to names.clj to provide that logic

Adding new test card that uses such a field reference

jeff303 committed:
Serialization fixes: accept only-db-ids option for dump

Accept options map for the dump command.  Currently, the only supported option is `:only-db-ids`, which is expected to be a set of DB IDs to dump.  If set, then only these databases (and their respective tables, fields, and segments) will be dumped to the dump-dir.  If not set, then the previous behavior takes effect (all DBs).

Update the load_test.clj to set this to be only the new temporary DB to avoid tramping on existing `sample-dataset` and other instances

jeff303 committed:
Serialization fixes: don't include personal collections of other users

Use the `select-collections` fn to filter only collections that are public, or owned bythe given user (by way of email), plus all collections with those as an ancestor

Updating test to assert a non-owned personal collection is not persisted (including nested), but all nested personal collections are

Adding new wrapped macro to also bind collection/*allow-deleting-personal-collections* true around the with-temp* call so that personal collections can legitimately be deleted (since they're temporary); the name of this macro was chosen to preserve the same amount of spacing within `with-world`

camsaul commented:
@jeff303 in #15971 you mentioned you "fixed" the serialization of field literal clauses in this PR -- I mentioned there that #15971 is expected behavior... was that change needed to fix something else?

paulrosenzweig commented on shared/test/metabase/shared/models/visualization_settings_test.cljc:
```diff
@@ -0,0 +1,51 @@
+    (let [f-qual-nm "/databases/MY_DB/tables/MY_TBL/fields/COL1"
+          f-id      42
+          col-nm    "Year"]
+      (doseq [[input-str expected] [[(fmt "[\"ref\",[\"field\",%d,null]]" f-id) {::mb.viz/field-id f-id}]
```
Two things:
- Does this handle other field ids in the options? e.g. `["field", 123, {"source-field": 456}]`
- Pivot tables (and maybe other places) have lists of field refs that aren't prefixed with "ref".

jeff303 replied:
I think I will need to gather these examples to beef up this test (even if they aren't immediately relevant for serialization).

paulrosenzweig commented on shared/src/metabase/shared/models/visualization_settings.cljc:
```diff
@@ -0,0 +1,265 @@
+  `db-normalized-entry-to-db-form-entry-to-normalized`."
+  [m k v]
+  (case k
+    :click_behavior (assoc m ::click-behavior (-> v
```
It's not clear to me from reading the code, are we handling click behavior defined on table columns or just click behavior defined on the top level?

jeff303 replied:
As of now, it's **only** column level click actions.  I think I need to see some examples of top level still...

paulrosenzweig reviewed (approved):
I didn't review the logic but left a few "did we handle X?" comments in the vis settings code. Looks good as far as I can see!

jeff303 committed:
Serialization fixes: support native query snippets

Adding support for NativeQuerySnippet:

- in names
- in dump (putting into subdirectory of collection, as cards)
- in upsert (identity condition of :name :collection_id)
- in load (no retry fn necessary since they should not have any dependencies themselves)

Adding a snippet and a card referencing it to roundtrip test (including a deeply nested snippet)

Fixing up handling of snippet namespaced collections and adding relevant name tests

jeff303 commented on shared/test/metabase/shared/models/visualization_settings_test.cljc:
```diff
@@ -0,0 +1,51 @@
+                               :linkType         "question"
+                               :parameterMapping {}
+                               :targetId         target-id}
+          db-col-settings     {(fmt "[\"ref\",[\"field\",%d,null]]" f-id) {:click_behavior db-click-behavior}
```
The serialization roundtrip does pass through an MBQL normalization layer that should do this.  That being said, we officially only support loading dumps created in the same version.  So in this case, we would only ever need support in .39 or later (which this PR targets), so as long as only the new forms ever appear, we should be OK.  Not sure if it's possible for an older card or something to be hanging around still with the older forms.

jeff303 commented:
> @jeff303 just wanted to make sure you saw my comment in #15971 -- dumping as `field` is correct behavior, so it's probably an issue with `load` handling those wrong

Yep, that was indeed the case.  Load wasn't correctly distinguishing between a column reference by name versus a fully qualified "column ID" in the context of serialization.

rlotun commented on enterprise/backend/src/metabase_enterprise/serialization/serialize.clj:
```diff
@@ -15,24 +15,27 @@
 
   This gets stored with each dump, so we can correctly recover old dumps."
-  1)
+  ;; version 2 - start adding namespace portion to /collections/ paths
```
we probably don't have a convention and official docs for the serialization protocol, but assuming this is strictly a superset and backwards compatible (?) then would 1.1 be more appropriate?

jeff303 replied:
This is a big enough change that I think a bump is warranted (even if it's not used anywhere).  And this field is a `Long` so we can't do decimals.

camsaul commented on enterprise/backend/src/metabase_enterprise/serialization/cmd.clj:
```diff
@@ -47,6 +48,7 @@
                        context)]
     (try
       (do
+        (log/infof "BEGIN LOAD from %s with context %s" path context)
```
You should probably i18n INFOs. For DEBUG and TRACE I think no i18n is fine since people don't really enable those unless someone (usually us) is debugging something

camsaul replied:
bump

jeff303 replied:
Sorry, forgot about this one.  Should be fixed now.

camsaul commented on enterprise/backend/src/metabase_enterprise/serialization/serialize.clj:
```diff
@@ -108,19 +129,74 @@
-        series          (when (not-empty dashboard-cards)
-                          (db/select DashboardCardSeries
-                            :dashboardcard_id [:in (map u/the-id dashboard-cards)]))]
+  (let [dashboard-cards   (db/select DashboardCard :dashboard_id (u/the-id dashboard))
```
Not sure if this indentation change was intentional

jeff303 replied:
I think this is fixed now, or at least I reverted that spacing change in the latest.

camsaul commented on enterprise/backend/src/metabase_enterprise/serialization/upsert.clj:
```diff
@@ -29,14 +30,14 @@
-   Collection          [:name :location]
+   Collection          [:name :location :namespace]
    Dashboard           [:name :collection_id]
-   DashboardCard       [:card_id :dashboard_id :visualiation_settings]
```
Maybe we should have some error handling that would warn if these keys don't exist? Either in tests or in runtime checks (that would cause test failures). To prevent this sort of stuff in the future

jeff303 replied:
Added a check to `upsert_test.clj` for this.

camsaul commented on shared/src/metabase/shared/models/visualization_settings.cljc:
```diff
@@ -0,0 +1,410 @@
+  #?(:clj  (json/encode obj)
+     :cljs (.stringify js/JSON (clj->js obj))))
+
+(defn parse-column-ref
+  "Opposite of the `column-ref-for-*` fns. Converts the given string representation of a column settings key into its
+  normalized form. The `col-ref` argument can be a string or keyword (which is produced by YAML parsing, for instance).
+  In case it is a keyword, it will be converted to its full name. \"Full\" means that the portions before and after any
+  slashes will be included verbatim (via the `keyname` helper fn). This is necessary because our serialization code
+  considers a forward slash to be a legitimate portion of a fully qualified name, whereas Clojure considers it to be a
+  namespace/local name separator."
+  {:added "0.40.0"}
+  [col-ref]
+  (let [parts           ((comp vec parse-json-string) (if (keyword? col-ref) (keyname col-ref) col-ref))
+        [kind & params] parts]
+    (case kind
+      "ref"  (let [[[t id & _]] params]
+               (case t "field" (cond (int? id)    {::field-id id}
+                                     (string? id) {::field-qualified-name id})))
+      "name" {::column-name (first params)})))
```
Why not just use Spec to do all of this. Seems like it would be a lot easier to write a spec for the shape and have it tell you what it got

jeff303 replied:
I have some specs defined but I'm unclear on what you mean here.  Do you mean using `conform`?

jeff303 replied:
@camsaul , can you clarify on this one?  Doing a big cleanup/PR feedback commit and I wanted to understand this better.

jeff303 replied:
OK, have a look at the changes now to see if that's what you had in mind.

camsaul commented on shared/src/metabase/shared/models/visualization_settings.cljc:
```diff
@@ -0,0 +1,410 @@
+  [settings col-key action]
+  (-> settings
+      with-col-settings
+      (update-in [::column-settings] #(assoc % col-key {::click-behavior action}))))
```
You don't need to use a function here. You can just do `(update-in ks assoc col-key ...)`

jeff303 replied:
Fixed.

camsaul commented on shared/src/metabase/shared/models/visualization_settings.cljc:
```diff
@@ -0,0 +1,410 @@
+  "Creates a parameter mapping for `source-col-name` (`source-field-id`) to `target-field-id` in normalized form."
+  {:added "0.40.0"}
+  [source-col-name source-field-id target-field-id]
+  (let [id         [:dimension [:fk-> [:field-id source-field-id] [:field-id target-field-id]]]
```
```clj
[:fk-> [:field-id source-field-id] [:field-id target-field-id]]
```

is legacy MBQL syntax, don't use it please 😬 

You want 

```clj
[:field target-field-id {:source-field source-field-id}]
```

instead

jeff303 replied:
Fixed.

camsaul commented on shared/test/metabase/shared/models/visualization_settings_test.cljc:
```diff
@@ -0,0 +1,152 @@
+    (let [from-id    101
+          to-id      294
+          card-id    19852
+          mapping-id (fmt "[\"dimension\",[\"fk->\",[\"field-id\",%d],[\"field-id\",%d]]]" from-id to-id)
```
Legacy MBQL syntax.

Run it thru the normalizer and let that take care of it. That should be the only place in the whole codebase that knows about or deals with legacy syntax

jeff303 replied:
Fixed.

camsaul commented on enterprise/backend/src/metabase_enterprise/serialization/cmd.clj:
```diff
@@ -103,29 +109,45 @@
-         tables (Table)
-         collections (Collection)] ;(select-collections users state)]
+         databases   (if (contains? opts :only-db-ids)
+                       (db/select Database :id [:in (:only-db-ids opts)])
+                       (Database))
```
It it worth ordering these by `:id` or something, just so serializations are repeatable? Otherwise the order stuff comes back in is sort of up to the database

jeff303 replied:
Changed to order them all by `:id`.

camsaul reviewed (approved):
This looks like a big improvement over the old serialization code. I'm a big fan of all the new viz settings stuff in the shared lib too.

My only request is to change the stuff using legacy MBQL syntax to use the new syntax. Ideally everyone in the whole world outside of the `normalize` namespace should forget `fk->` ever existed

jeff303 committed:
Serialization fixes: handle source-table for joins

Adding `mbql-fully-qualified-names->ids*` for recursive calls, which does not normalize its args, keeping mbql-fully-qualified-names->ids` as a wrapper to that

Adding clause to `mbql-fully-qualified-names->ids*` to resolve `:source-table` when it's a fully qualified table name

Adding util fn to names.clj to support that scenario

Updating "My Card" test card to include a join

Skipping two more drivers from `dump-load-entities-test` - SQLite and Spark because they don't support FKs

jeff303 committed:
Serialization fixes: cleanup and refactoring

Responding to PR feedback

Updating upsert_test.clj to check that the identity-condition keys actually exist for the models

Removing legacy MBQL syntax in a couple places

Adding ordering by `:id` to serialization load command for predictability

Lots of fixes and refactorings in visualization_settings.cljc:
 - added lots of specs and fdefs
 - added test fixture that instruments all the speced fns and required it from both visualization_settings.cljc and load_test.clj
 - renamed a bunch of fns for more consistency
 - added and fixed some docstrings
 - fix implementation of `parse-json-string` for cljs

flamber reviewed (commented):
Awesome journey so far! :tada:
These are last problems I see (tested `bffb807`).
1. #15120 - still dumping PC, but load does not include them anymore. But the problem is that it now loads PC sub-collections in root, which can be even more problematic because of the naming schema. If we cannot fix the issue, then let's at least revert back to just including the PC, since those are more manageable.
2. New on this PR - Field dimension `temporal-unit` not included in `dump` for implicit joined aggregation, which means `load` will use Day as aggregation - can lead to Correctness issue (at least P2)
Simple question > Sample Dataset > Orders - summarize Count by User.CreatedAt:Year - save question - dump - load.
3. #15283 - it doesn't handle joins to Saved Questions. It seems like it's "just" not doing translation on `load` - likely because it's referencing a question, but perhaps the code is expecting table reference?

jeff303 committed:
Serialization fixes: preserve temporal-unit in source-field clause

Update serialize and match macros to keep the rest of the match, when modifying the source field ID

Adding a bunch of new tables in order to perform a "bucketing on implicit join" type query

Updating test_util.clj to initialize the query processor with all the new temp tables so that queries actually work

jeff303 commented:
Commit 15 should fix issue 2 from @flamber latest comment.  Will try to address the other two points tomorrow.

jeff303 committed:
Serialization fixes: misc cleanup and fixes

Fixing the implementation of `fully-qualified-name->context` to work properly when the entity name itself is also a model name

jeff303 commented:
Commit 11 has been reworked to, I believe, address the personal collections problem.  In particular:

- If `--user` is specified, then that user's PC (and any sub collections, either direct or indirect) will be serialized (along with public collections having no `personal_owner_id`, of course)
- PCs for other users (and their direct or indirect sub-collections) will not be serialized
- Some tests added to ensure that the sub-collections are deserialized properly (i.e. have location set correctly)

camsaul reviewed (approved):
I didn't go over this is a ton of detail this time since it looks the same overall approach as last time I reviewed it... giving it another 👍

flamber reviewed (approved):
My brain can't anymore. Everything looks good, beside point 3 in https://github.com/metabase/metabase/pull/15858#pullrequestreview-663691231, which I think we should create a separate issue for, which will not be fixed as part of this PR.

jeff303 committed:
Serialization fixes: handle joining to cards

Recognize :source-table pointing to a fully qualified card name in `mbql-fully-qualified-names->ids*`, and adding the unresolved name capture to that lookup

Adding new util fns to capture the unresolved named entries from anywhere within the MBQL query tree: `paths-to-key-in` and `gather-all-unresolved-names`

Updating `resolve-card-dataset-query` to call this new `gather-all-unresolved-names` to pull all unresolved names from the MBQL query tree into card itself

Renaming `fully-qualified-name->id-rec` to `resolve-source-query` for clarity, and removing now unneeded clause

To the test case, adding a card that joins to another card

flamber commented:
Awesome work :heart: - did a full test through everything this fixes the three points I had.

----

Though noticed a problem, which didn't occur in an earlier commit (have no idea, when it changed).

There's a Click Behavior created on 38, which has columns that looks like this on the SOURCE report_dashboardcard.visualization_settings.column_settings:
`... "[\"dimension\",[\"fk->\",[\"field-id\",13],[\"field-id\",4]]]" ...`

When doing Serialization on latest commit, it does this during dump:
`... '["dimension",["fk->",["field-id","/databases/Sample Dataset-source/schemas/PUBLIC/tables/ORDERS/fks/PRODUCT_ID"],["field-id","/databases/Sample Dataset-source/schemas/PUBLIC/tables/PRODUCTS/fields/CATEGORY"]]]' ...`

On load, the TARGET report_dashboardcard.visualization_settings.column_settings looks like this:
`... "[\"dimension\",[\"fk->\",[\"field\",1487,null],[\"field\",1472,null]]]" ...`

But that is a wrong dimension format generated during dump/load - the final result should look like this:
`... "[\"dimension\",[\"field\",1472,{\"source-field\":1487}]]" ...`

jeff303 commented:
I could probably fix this in a follow-on patch, but I'm a bit out of my depth on the MBQL details.  @camsaul , can you comment here?  Which is the "new" syntax, and what MBQL function would be responsible for doing that translation?  It seems that this is a no-op.

```
(mbql.normalize/normalize-tokens [:dimension [:fk-> [:field-id 1] [:field-id 2]]])
=> [:dimension [:fk-> [:field-id 1] [:field-id 2]]]
```

flamber commented:
I have no idea how, but on 1.39.3 it works correctly in the frontend anymore, while it still creates this wrong-looking dimension.
Not going to create an issue for it then, since I cannot reproduce the problem on release, and would likely only be a problem for pre-39 created dashboards with a very specific way to apply Click Behavior.
